pennylane/labs/vibrational_ham/localize_modes.py

```diff
@@ -0,0 +1,206 @@
+    described in this `work <https://pubs.aip.org/aip/jcp/article-abstract/141/10/104105/74317/Efficient-anharmonic-vibrational-spectroscopy-for?redirectedFrom=fulltext>`_
+    """
+    nmodes = len(freqs)
+    hess_normal = np.zeros((nmodes, nmodes))
+    for m in range(nmodes):
+        hess_normal[m, m] = freqs[m] ** 2
+
+    natoms, _ = np.shape(disp_vecs[0])
+
+    qmat = np.zeros((natoms, 3, nmodes))
+    for m in range(nmodes):
+        dvec = disp_vecs[m]
+        for i in range(natoms):
+            for alpha in range(3):
+                qmat[i, alpha, m] = dvec[i, alpha]
+
+    uloc, qloc = _localization_unitary(qmat)
+    hess_loc = uloc.transpose() @ hess_normal @ uloc
+    loc_freqs = np.sqrt(np.array([hess_loc[m, m] for m in range(nmodes)]))
+
+    if order is True:
+        loc_perm = np.argsort(loc_freqs)
+        loc_freqs = loc_freqs[loc_perm]
+        qloc = qloc[:, :, loc_perm]
+        uloc = uloc[:, loc_perm]
+
+    return loc_freqs, qloc, uloc
+
+
+def localize_normal_modes(results, freq_separation=[2600]):
+    """
+    Arguments: results dictionary obtained from harmonic_analysis
+    separates frequencies at each point in freq_separation array
+
+    Returns: new dictionary with information for localized modes
+    """
+    freqs_in_cm = results["freq_wavenumber"]
+    freqs = freqs_in_cm / au_to_cm
+    disp_vecs = results["norm_mode"]
+    nmodes = len(freqs)
+
+    modes_arr = []
+    freqs_arr = []
+    disps_arr = []
+    num_seps = len(freq_separation)
+    min_modes = np.nonzero(freqs_in_cm <= freq_separation[0])[0]
+
+    modes_arr.append(min_modes)
+    freqs_arr.append(freqs[min_modes])
+    disps_arr.append(disp_vecs[min_modes])
+
+    for i_sep in range(num_seps - 1):
+        is_bigger = np.array(freq_separation[i_sep] <= freqs_in_cm)
+        is_smaller = np.array(freq_separation[i_sep + 1] >= freqs_in_cm)
+        mid_modes = np.nonzero(is_bigger * is_smaller)[0]
+        modes_arr.append(mid_modes)
+        freqs_arr.append(freqs[mid_modes])
+        disps_arr.append(disp_vecs[mid_modes])
+
+    max_modes = np.nonzero(freqs_in_cm >= freq_separation[-1])[0]
+
+    modes_arr.append(max_modes)
+    freqs_arr.append(freqs[max_modes])
+    disps_arr.append(disp_vecs[max_modes])
+
+    natoms = np.shape(disp_vecs[0])[0]
+
+    loc_freqs_arr = []
+    qlocs_arr = []
+    ulocs_arr = []
+    for idx in range(num_seps + 1):
+        num_freqs = len(freqs_arr[idx])
+        loc_freqs = []
+        uloc = []
+        qloc = []
+        if num_freqs > 1:
+            loc_freqs, qloc, uloc = _localize_modes(freqs_arr[idx], disps_arr[idx])
+        elif num_freqs == 1:
+            loc_freqs = freqs_arr[idx]
+            qloc = np.zeros((natoms, 3, 1))
+            qloc[:, :, 0] = disps_arr[idx][0]
+            uloc = np.zeros((1, 1))
+            uloc[0, 0] = 1
+
+        loc_freqs_arr.append(loc_freqs)
+        qlocs_arr.append(qloc)
+        ulocs_arr.append(uloc)
+
+    uloc = np.zeros((nmodes, nmodes))
```

Comment:
Im not a huge fan of having uloc being redefined when it was used previously in a loop as an inner variable. The one inside the for loop should change its name so that there isn't an overlap imo. A less careful reader might see it returns uloc at the bottom and assume the uloc in line 172 is what's eventually returned.